Privacy radio button

to6ko12 opened the pull request:
[sc-27471] [sc-27475] [sc-27476]

to6ko12 committed:
Add radio button to settings

to6ko12 committed:
Add json snippet in thank you page

to6ko12 committed:
Remove hardcoded value

to6ko12 committed:
First draft for consent API call
should be tested and refined when 'has_permission' API endpoint is completed

to6ko12 committed:
Change code to connect to API endpoint

to6ko12 committed:
Readability improvements

TsankoBg commented on common/src/API.php:
```diff
@@ -63,6 +63,21 @@ private function buildURL($address, $parameters) {
+                'https://%s/api/order_permissions?order_number=%u&id=%u',
+                $this->api_domain,
+                $data['order'],
+                $this->shop_id
```
Why are you adding the `order_number` and `id` to the query params with `sprintf`, and adding again `id` and `code` via `buildURL` `parameters`. This probably adds `id` twice to the query ...pick one way, and try to be consistent with the rest of the code.

TsankoBg commented on common/src/API.php:
```diff
@@ -63,6 +63,21 @@ private function buildURL($address, $parameters) {
+            ),
+            ['id' => $this->shop_id, 'code' => $this->api_key]
+        );
+        $permission_response = Requests::post($permission_url, []);
```
Did you even test this? 
The endpoint is for a `GET` request, and you are doing a `POST` request here, this probably does not work.
https://github.com/webwinkelkeur/dash/pull/2059/files 

You should test your code before submitting a PR.

TsankoBg commented on common/src/WooCommerce.php:
```diff
@@ -148,6 +149,10 @@ private function sendInvite($order_id) {
 
         // send invite
         $api = new API($api_domain, $shop_id, $api_key);
+        if ($this->plugin->getOption('invite') == 3 && !$api->hasConsent($data)) {
```
Did you test what happens in case of an exception or bad request data, e.g. empty `order_number`, wrong or empty credentials

TsankoBg commented on common/src/WooCommerce.php:
```diff
@@ -589,4 +594,23 @@ private function isProductReviewsEnabled(): bool {
         return strtotime($this->getLastReviewSync()) > strtotime('-24 hours');
     }
+
+    public function addOrderDataJsonThankYouPage() {
```
Avoid writing code so nested.

TsankoBg commented on common/src/WooCommerce.php:
```diff
@@ -589,4 +594,23 @@ private function isProductReviewsEnabled(): bool {
+
+    public function addOrderDataJsonThankYouPage() {
+        if (is_wc_endpoint_url('order-received')) {
+            if ($this->plugin->getOption('invite') == 3) {
```
Let's improve this a bit. It's hard to know which option is 3, would be nice to use a `const` here.

TsankoBg commented on common/src/WooCommerce.php:
```diff
@@ -25,6 +25,7 @@ public function __construct(BasePlugin $plugin) {
         add_action($this->getReviewsHook(), [$this, 'syncReviews']);
         add_action('wp_ajax_' . $this->getManualSyncAction(), [$this, 'manualReviewSync']);
         add_action('wp_ajax_' . $this->getProductKeysAction(), [$this, 'getProductKeys']);
+        add_action('wp_head', [$this, 'addOrderDataJsonThankYouPage']);
```
It might be worth checking `woocommerce_before_thankyou`, `wp_head` will be fired on every page.

You have to also check if the order did not fail.

to6ko12 committed:
Connect to API 2.0

to6ko12 committed:
Monitor thank you page, not wp_head

TsankoBg commented on common/src/API.php:
```diff
@@ -63,6 +63,26 @@ private function buildURL($address, $parameters) {
+            'code' => $this->api_key,
+        ];
+        $permission_url = $this->buildURL('https://' . $this->api_domain . '/api/2.0/order_permissions.json', $params);
+        $permission_response = Requests::get($permission_url, []);
```
```suggestion
        $permission_response = Requests::get($permission_url);
```

No need for the second parameter? https://developer.wordpress.org/reference/classes/requests/get/

TsankoBg replied:
I think this is a deprecated library, could you also check that?

TsankoBg commented on common/src/API.php:
```diff
@@ -63,6 +63,26 @@ private function buildURL($address, $parameters) {
+        if ($response_code != 200) {
+            throw new WebwinkelKeurAPIError(
+                $permission_url,
+                $permission->message ?? $permission->body
```
I would suggest adding some hard-coded message instead of throwing the whole response body
` $permission->message ?? 'Invitiation could not be created due to .....'`

TsankoBg commented on common/src/WooCommerce.php:
```diff
@@ -589,4 +610,23 @@ private function isProductReviewsEnabled(): bool {
+        if (!is_wc_endpoint_url('order-received') || $this->plugin->getOption('invite') != self::POPUP_OPTION) {
+            return;
+        }
+        $order_id = absint(get_query_var(get_option('woocommerce_checkout_order_received_endpoint')));
```
You can get the order_id passed as a parameter to this function instead of retrieving it
```
public function addOrderDataJsonThankYouPage(int $order_id)
```

TsankoBg commented on common/src/WooCommerce.php:
```diff
@@ -589,4 +610,23 @@ private function isProductReviewsEnabled(): bool {
     }
+
+    public function addOrderDataJsonThankYouPage() {
+        if (!is_wc_endpoint_url('order-received') || $this->plugin->getOption('invite') != self::POPUP_OPTION) {
```
No need for this check because you are already on this page?
`!is_wc_endpoint_url('order-received')`

TsankoBg commented on common/src/WooCommerce.php:
```diff
@@ -148,6 +153,22 @@ private function sendInvite($order_id) {
+                sprintf(
+                    __('The %s invitation could not be sent.', 'webwinkelkeur'),
+                    $this->plugin->getName()
+                ) . ' ' . $e->getMessage()
```
Instead of `sprintf` and then `.  '  '  .`, just put everything inside the `sprintf`

TsankoBg reviewed (changes requested):
Minor things to be checked

to6ko12 committed:
Small fixes
1. stop using deprecated library
2. fix readability
3. add translations

TsankoBg commented on common/src/API.php:
```diff
@@ -63,6 +63,26 @@ private function buildURL($address, $parameters) {
+        if ($response_code != 200) {
+            throw new WebwinkelKeurAPIError(
+                $permission_url,
+                $permission->message ?? 'There was an error sending the invitation to the dashboard.'
```
```suggestion
                $permission->message ?? 'There was an error retrieving the order consent.'
```

TsankoBg reviewed (changes requested):
Looks good, only the message should be adjusted.

to6ko12 committed:
Update common/src/API.php

Co-authored-by: Tsanko Hadzhiev <tsankohadzhiev@gmail.com>

TsankoBg committed:
orderId -> orderNumber
Everywhere else we call it `orderNumber`

TsankoBg committed:
`woocommerce_before_thankyou` -> `wp_head`
`woocommerce_before_thankyou` does not get triggered on live test shop. Will keep `wp_head` until I find a better solution

TsankoBg committed:
order_number -> orderNumber

TsankoBg committed:
Add inviteDelay to the json order data

TsankoBg committed:
Add Hash.php

TsankoBg committed:
Add signature to order data

TsankoBg committed:
Add empty line to the end of Hash.php

TsankoBg committed:
Formatting addOrderDataJson

TsankoBg committed:
Move InvalidKeysException to its own file

TsankoBg committed:
Fix exception

TsankoBg committed:
Hash code to php 7.4

TsankoBg committed:
Add webshopId to order data

TsankoBg commented on common/src/API.php:
```diff
@@ -63,6 +63,26 @@ private function buildURL($address, $parameters) {
         return $address . '&' . $query_string;
     }
+
+    public function hasConsent(array $data): bool {
```
`$data` can be `$order_number`
```suggestion
    public function hasConsent(array $order_number): bool {
```

TsankoBg reviewed (changes requested):
Looks good, only 1 minor thing has to be updated in `hasConsent`

TsankoBg committed:
Pass order_number directly to hasConsent

TsankoBg committed:
Revert Requests changes
To be updated and tested in future version

TsankoBg committed:
FIx typo
